Implement pruning in base trainer

dorimedini opened the pull request:
Implements #83 

@galshachaf something tells me the specific implementation of the pruning process depends on the network architecture more than I took into account... specifically, maybe the `_prune` method should raise `NotImplementedError` in `BaseTrainer` and I should move the implementation to `FCTrainer`?

dorimedini committed:
go() now has extra post-train step

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Now allow pruning (by default) in BaseTrainer

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Added setter / getter

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Export pruning to static method so we can use it with external models

Signed-off-by: dorimedini <dorimedini@gmail.com>

noamloya commented on trainer.py:
```diff
@@ -92,12 +101,52 @@ def get_epochs(self):
+            input_weights = weights[0]  # weights[1] is the list of node biases
+            # The incoming edge weights of node N is incoming_edge_weights[N]
+            new_weights = []
+            for node_input_weights in input_weights:
```
I think this would run faster if it will be implemented as numpy matrix operations:

`w = np.array(weights)`
`new_weights = np.where(w<threshold, 0, w)`

or something like that

dorimedini replied:
Looks good, but how would I compute `pruned_edges` though in a numpy way? ATM I'm just iterating

noamloya replied:
Use [numpy.count_nonzero](https://docs.scipy.org/doc/numpy/reference/generated/numpy.count_nonzero.html) :)

dorimedini replied:
I'm having a problem using your suggestion... `np.where(w<threshold, 0, w)` isn't good enough because if `w` is negative with large absolute value it will still be nullified. But `np.where(math.fabs(w)<threshold, 0, w)` complains when I try to do `math.fabs(w)` where `w` is not a 1-dimensional array.
I can do:
```
w = np.array(weights)
w_abs = np.absolute(w)
new_weights = np.where(w_abs<threshold, 0, w_abs)
```
...but the `np.absolute(w)` call seems expensive. Do you think it's worth it?

dorimedini replied:
nvm the answer is `(w<threshold) & (w>-threshold)`

galshachaf commented:
use get_weighted_layers_indices to get the specific layers that have weights.

galshachaf commented on trainer.py:
```diff
@@ -92,12 +101,52 @@ def get_epochs(self):
+            return
+        # Layer 0 has no input edges, start from layer 1
+        pruned_edges = 0
+        for i in range(1, len(model.layers)):
```
use use get_weighted_layers_indices to get the specific layers that have weights to get the list of layers that have weights.

galshachaf reviewed (changes requested):
use get_weighted_layers_indices to get the specific layers that have weights.

dorimedini committed:
Moved pruning to FCTrainer

Base trainer shouldn't know how to implement this, it may
depend on the type of NN

Signed-off-by: dorimedini <dorimedini@gmail.com>

dorimedini committed:
Optimized pruning

Signed-off-by: dorimedini <dorimedini@gmail.com>
